### ISSUE TITLE
qwen3: batched decode is not batch-invariant — a request's tokens depend on its batch-mates

### ISSUE THREAD

**Issue by xiaguan** [xiaguan/pegainfer#414]
## Problem

openinfer's qwen3 batched decode is **not batch-invariant**: the logits a request sees depend on *which other requests share its batch*. Same prompt, same weights, same GPU — but decoded alone vs. co-batched, a request can pick a different token at a near-tie and from there generate an entirely different sequence.

This is benign under the current correctness contract (greedy + the tolerance-based golden gate), but it means **a request's output is a function of the scheduler's batching decisions, not just its own input**. This issue is to (a) pin down *exactly which kernels* break invariance and (b) decide whether batch invariance is a property we want to guarantee.

## Evidence

Minimal repro (two prompts, greedy, prefill only):

- `prompt_a = [9707]` prefilled **alone** → first token = **1**
- `prompt_a = [9707]` prefilled **batched** with a second prompt → first token = **27**

The divergence happens *at prefill*, before decode starts. `prompt_a`'s last-position logits are a near-tie between token 1 and token 27; batching shifts them by ~1 bf16 ULP and flips the argmax. After that the two runs decode different token streams and diverge arbitrarily (this is why a naive per-step logit diff looks "gross" — it is comparing two unrelated generations).

The team already documented the mechanism. From `openinfer-qwen3-4b/tests/hf_golden_gate.rs`:

> batched eager ... replaces the old exact `batch==sequential` check, which mistook the batched decode path's benign reduction-order noise (within tolerance here) for a bug — **batch composition changes the reduction order and drifts logits ~1 ULP.**

The model-gated `batch_matches_sequential` exact check — which asserted batched decode == per-request sequential decode token-for-token — is what surfaces this. It is superseded by the tolerance gate and slated for removal; it is **not** a real engine regression, it is the wrong correctness contract for a bf16 + cuBLAS engine.

## Root cause (so far)

**Primary suspect: cuBLAS/cuBLASLt GEMM algorithm selection is keyed on N (the token/batch dimension).** From `openinfer-kernels/src/ops/linear.rs:7`:

> GEMMs at or below this N consult the cublasLt plan cache. 32 covers the decode buckets where cuBLAS's GemmEx heuristic picks badly (worst at N in [8, 16], where it **skips split-K entirely**); above that the GEMMs are wide enough that the default selection is fine.

For `N <= 32` the algo is tuned per `(num_rows, n, cols)` shape (`gemm_lt_tune`); above it, cuBLAS's default heuristic is also shape-dependent. Either way the chosen algorithm — and its split-K factor — is a function of N. Two batched prefills at different total token counts (`prompt_a` alone vs. `prompt_a` + `prompt_b`) therefore hit the heuristic at different N, can land on different algorithms / split-K, and produce ~1 ULP-different logits for `prompt_a`'s shared rows. This is the classic batch-invariance failure mode.

It also matches the prefill evidence: in varlen prefill `prompt_a`'s attention is over its own KV regardless of batch-mates, so attention is *not* the obvious culprit there — the GEMMs (QKV / O / MLP / lm_head), whose N grows when a second prompt joins, are.

**Secondary suspects to rule in/out:**
- **lm_head GEMM** — `batch_token_logits` (`openinfer-qwen3-4b/src/prefill.rs:289`), same cuBLAS story, directly produces the logits that get argmaxed.
- **Decode-time split-KV attention** — `paged_attention_batch_decode_split_kv_into` (`openinfer-core/src/ops/attention.rs`) splits KV; if the split count depends on batch shape, the softmax-rescale reduction order moves with batch composition (same class as the Marlin split-K / PPLX combine wobble in #286).
- **RMSNorm / elementwise** — per-row, expected batch-invariant; confirm and exclude.

## Why it matters / relation to existing issues

This is the **batch-composition axis** of the near-tie family we already track:
- #286 (kimi) — near-tie argmax wobbles *run-to-run* with a fixed batch (atomic-reduction order: Marlin split-K / PPLX combine).
- #384 (qwen3 lora) — a bf16 tie breaks differently *across allocations* on multi-GPU.

Both are reduction-order coin flips with the batch held fixed. This issue is the orthogonal axis: **hold the request fixed, vary the batch, and the output moves.** In production continuous batching that means a request's generated tokens depend on whoever happened to be co-scheduled at each step — runs are not reproducible across batch schedules, and a request replayed under different load can answer differently.

## What to investigate ("why, specifically")

1. **Localize the first divergent op.** Decode `prompt_a` alone vs. co-batched, dumping intermediate activations layer-by-layer (norm → QKV → attn → O → MLP → lm_head); identify the first op whose output for `prompt_a`'s rows differs, to confirm GEMM N-dependence is dominant (and quantify any attention contribution).
2. **Quantify the blast radius.** On a real workload, how often is a position near-tied tightly enough (within a few bf16 ULP — the golden gate's `MARGIN_TOL` regime) that batch composition flips the pick?
3. **Decide scope.** Is "stable within bf16 tolerance vs. HF" (the current gate) the contract, or do we want batch invariance as a guarantee? Batch-invariant kernels cost throughput on narrow batches.

## Prior art

- Thinking Machines Lab, *Defeating Nondeterminism in LLM Inference* (2025) — identifies non-batch-invariant matmul / attention / norm reductions as the root of inference nondeterminism and gives batch-invariant kernel implementations.
- vLLM ships a `batch_invariant` mode built on the same idea.

If we pursue this, the fix direction is a batch-invariant GEMM (fixed algorithm + fixed split-K regardless of N) and a fixed-split attention reduction, accepting the throughput cost on narrow batches.

## Scope note

This is **not a regression and not a correctness bug** under the current contract — the tolerance golden gate (`hf_golden_gate`) passes and is the right gate. This issue is about *understanding the exact non-invariance sources* and *deciding* whether to guarantee batch invariance. Tracking under the qwen3 line (#241).

**Comment by FeathBow** [comment]
I'd like to paste some findings here.

## Evidence (Q1)

On Single GPU (x86_64, sm_89), the first op whose output for a request's rows changes with batch composition is the **layer-0 Q projection GEMM** — and the same mechanism applies to every projection GEMM (QKV / O / gate-up / down, and lm_head). Batched prefill concatenates all prompts' tokens, so these run at `N = total batched tokens`. The dispatch (`launch_gemm` in `openinfer-kernels/src/ops/linear.rs`) selects a cuBLAS/cuBLASLt algorithm whose plan cache is keyed on `{M, N, K}`, so **N selects the algorithm, and with it the split-K factor.** Under bf16 inputs with fp32 accumulation, a different split-K factor means a different reduction order, which shifts the shared rows by ~1 ULP may enough to flip a near-tie argmax.

Two-sided evidence on Single GPU (x86_64, sm_89). First, the observational half — a per-op fingerprint (token 0, prefilled alone vs co-batched): embedding and the pre-GEMM RMSNorm are bit-identical (per-row), and the **first fingerprint to change is the Q-proj GEMM**. (In the logs below: `fnv` = a hash of the whole op-output vector, `sum` = its sum; `lp0` = logprob of the alone-run's argmax token; `Nb` = filler tokens added as a batch-mate; `total_N` = total tokens in the batched GEMM; `maxΔlp_vs_alone` = max logprob change vs the alone run.)


<details><summary>Per-op dump — embedding & RMSNorm identical; first divergence at the Q-proj GEMM</summary>

```
===== ALONE (prompt_a=[9707]) =====
L0.in_embedding      sum=-0.967137 fnv=2fad72b4ba6aeb25
L0.post_input_norm   sum=-1.413508 fnv=f24d1b7beae1eb25
L0.post_Q_proj_gemm  sum=-2.184165 fnv=9758f7f187c36325
L0.post_attention    sum= 5.913832 fnv=a6b83865d8c56325
L0.out_residual      sum= 3.782215 fnv=de6d93fb68fceb25
===== BATCHED (prompt_a=[9707] + filler [785; 15]) =====
L0.in_embedding      sum=-0.967137 fnv=2fad72b4ba6aeb25   # identical
L0.post_input_norm   sum=-1.413508 fnv=f24d1b7beae1eb25   # identical
L0.post_Q_proj_gemm  sum=-2.182944 fnv=14e12ee1c7dc6325   # <-- FIRST DIVERGENCE
L0.post_attention    sum= 5.926523 fnv=38a96ae828996325   # downstream of Q (expected)
L0.out_residual      sum= 3.817169 fnv=ad4cb283912eeb25   # downstream of Q (expected)
```

Values agree to ~5 digits while the whole-vector FNV differs — a ~1-ULP spread. Attention diverges only *after* Q; in prefill it is per-request causal, so it inherits the perturbed Q rather than causing it.
</details>


Second, the **causal half** — force every projection to a fixed N=1 per-token GEMM and the divergence vanishes to bit-exactness (`max|Δlogprob| = 0` at every batch width):

<details><summary>Pin the GEMM to fixed N=1 → batch composition produces bit-identical logits</summary>

```
ALONE   argmax=1 lp0=-3.35660 det_rerun_bit_identical=true
BATCH +Nb=1  total_N=2  argmax=1 lp0=-3.35660 maxΔlp_vs_alone=0.000000 argmax_flip=no
BATCH +Nb=3  total_N=4  argmax=1 lp0=-3.35660 maxΔlp_vs_alone=0.000000 argmax_flip=no
BATCH +Nb=7  total_N=8  argmax=1 lp0=-3.35660 maxΔlp_vs_alone=0.000000 argmax_flip=no
BATCH +Nb=15 total_N=16 argmax=1 lp0=-3.35660 maxΔlp_vs_alone=0.000000 argmax_flip=no
BATCH +Nb=31 total_N=32 argmax=1 lp0=-3.35660 maxΔlp_vs_alone=0.000000 argmax_flip=no
BATCH +Nb=63 total_N=64 argmax=1 lp0=-3.35660 maxΔlp_vs_alone=0.000000 argmax_flip=no
```

Remove the variable the diagnosis blames (N-dependent algo selection) → the whole effect goes to `Δ = 0`. This turns "correlated with N" into "caused by N".
</details>

The issue's exact repro reproduces — `[9707]` alone → token **1**, co-batched → **27** (its top-1/top-2 are an exact tie, so any perturbation flips it) — and the drift is universal across prompts:

<details><summary>Exact 1↔27 repro + universal drift (sm_89)</summary>


```
[blast] 64 prompts @ total_N=8: moved(Δlp>0)=64/64  argmax_flips=1  near_ties(gap<0.02)=4
        maxΔlp: mean=0.146 p99=0.575 max=0.575
ALONE        argmax=1  gap(top1-top2)=0.00000  det_rerun_bit_identical=true
BATCH +Nb=1  total_N=2   argmax=27 maxΔlp=0.167449 argmax_flip=YES
BATCH +Nb=15 total_N=16  argmax=27 maxΔlp=0.071297 argmax_flip=YES
+Nb=3/7/31/63 (remaining widths)   argmax=1   drift present, no flip
```
</details>

## Hardware (re: GH200)

Mechanism is hardware-general; the *manifestation* is HW/cuBLAS-version specific. The non-invariance **reproduces on GH200 (aarch64, sm_90)** — same 1↔27 near-tie *pair* (sm_90's alone-pick is 27, at a 0.0625-nat gap vs sm_89's dead 0.0000 tie), batch composition flips it, 64/64 move:

<details><summary>Reproduces on a second GPU arch — GH200 / sm_90 (single-case + blast)</summary>

```
[blast] 64 prompts @ total_N=8: moved(Δlp>0)=64/64  argmax_flips=1  near_ties(gap<0.02)=2
        maxΔlp: mean=0.091 p99=0.367 max=0.367
ALONE        argmax=27 gap(top1-top2)=0.06250 det_rerun_bit_identical=true
BATCH +Nb=1  total_N=2   argmax=1  maxΔlp=0.063267 argmax_flip=YES
BATCH +Nb=7  total_N=8   argmax=27 maxΔlp=0.000000 argmax_flip=no
BATCH +Nb=15 total_N=16  argmax=27 maxΔlp=0.000000 argmax_flip=no
BATCH +Nb=31 total_N=32  argmax=1  maxΔlp=0.132998 argmax_flip=YES
BATCH +Nb=63 total_N=64  argmax=1  maxΔlp=0.130493 argmax_flip=YES
```
</details>

*Which* N drifts differs by GPU (sm_89 flips at N=16; sm_90 does **not** at N=16 — `maxΔlp=0.000000` — but flips at 2/4/32/64), with smaller magnitude on Hopper. **Which op diverges first is also hardware-specific:** on GH200, embedding, RMSNorm, the Q proj, and attention are all bit-identical alone-vs-batched; only the layer output changes — so the first diverging op there is a *later* projection GEMM (O or MLP), not Q:

<details><summary>GH200 per-op dump — Q proj & attention identical; first divergence is a later projection GEMM (O/MLP), not Q</summary>

```
ALONE                 L0.post_Q_proj_gemm=14e12ee1c7dc6325  post_attention=4b2479ee8b9f6325  out_residual=93cc5ebf1895eb25
BATCHED +Nb=1  (N=2)  L0.post_Q_proj_gemm=14e12ee1c7dc6325  post_attention=4b2479ee8b9f6325  out_residual=81a7c0651b60eb25
BATCHED +Nb=15 (N=16) L0.post_Q_proj_gemm=14e12ee1c7dc6325  post_attention=4b2479ee8b9f6325  out_residual=81a7c0651b60eb25
BATCHED +Nb=63 (N=64) L0.post_Q_proj_gemm=14e12ee1c7dc6325  post_attention=4b2479ee8b9f6325  out_residual=8eda98467ddaeb25
```

(`in_embedding` + `post_input_norm` also identical across all rows, omitted.) Q proj and attention are bit-identical at every N; only the layer output changes → the diverging op is a later projection GEMM (O-proj or MLP), not Q.
</details>

The invariant across both archs: the source is  a **projection GEMM** (norm/embedding/attention stay bit-identical), and *which* GEMM is first is what the GPU/cuBLAS picks. On sm_90 attention is bit-identical at every N — so its contribution is zero, which answers that part of Q1. TP / multi-GPU does not remove it either (the sharded projection GEMMs are still N-keyed; the existing TP=2 golden gate should keep passing within bf16 tolerance).

**Comment by FeathBow** [comment]
I'll do some more testing on this and see how it goes.

**Comment by FeathBow** [comment]
Closing the loop on this issue's own asks, now that the engineering has moved through #435 and the merged PR stack (#428, #438, #449, #458, #462, #473).

## Done

- Localize the sources: the batch-composition dependence decomposes into four concrete axes, each pinned down in code and either fixed or explicitly tracked — projection-GEMM algo selection keyed on N (the primary suspect here, confirmed, fixed by #428's per-`{M,K}` cuBLASLt algo pin, made fail-loud with no per-token fallback in #462), split-KV chunk count following live batch shape (confirmed, fixed by #438/#449/#473), attention-path selection and the decode-vs-unified forward residual (both confirmed still open, tracked in #435 with the 0.2.0 plan).
- lm_head: ruled in and covered — `batch_token_logits` runs through the same shared GEMM seam as the projections, and the envelope test spans prefill/decode/unified shapes across the production N sweep; a passing run proves the pin served each with no fallback (fallback is a fail-loud bail, not a silent path, so a green run is the proof).
- The scope decision: batch invariance is pursued as an opt-in guarantee behind `--batch-invariant` (`NumericPolicy::Pin`) rather than a default — today that flag covers the two reduction-order axes (projection-GEMM, split-KV chunk-count) and explicitly not the two path-selection residuals, which are tracked in #435; the default stays Tuned + the tolerance golden gate, exactly the contract this issue's scope note anticipated. The original repro — `prompt_a` prefilled alone vs co-batched flipping its first token — is closed under the flag and gated at output level by `batch_invariance_endtoend`.

## WIP

- RMSNorm / elementwise "confirm and exclude": nothing landed anywhere (no test, no doc). Small; folding into the 0.2.0 work scoped in #435.

## Minor (not blocking)

- Blast radius on a real workload: never quantified, and the decision it was meant to inform has since been made. Proposal: either fold a near-tie-frequency measurement into the 0.2.0 release notes as the "what does the default mode actually risk" number, or explicitly drop the item — your call.

Remaining engineering (the two open selection axes, the output-level e2e bar, milestone scoping) is tracked in #435 as the single source of truth — status note posted there.

**Comment by FeathBow** [comment]
Closing this as the main task and follow-up items are completed.
